src/Proto.Actor/Futures.cs

```diff
@@ -1,33 +1,39 @@
-﻿// -----------------------------------------------------------------------
-//  <copyright file="Futures.cs" company="Asynkron HB">
-//      Copyright (C) 2015-2017 Asynkron HB All rights reserved
-//  </copyright>
-// -----------------------------------------------------------------------
-
-using System.Threading.Tasks;
-
-namespace Proto
-{
-    public class FutureProcess<T> : Process
-    {
-        private readonly TaskCompletionSource<T> _tcs;
-
-        public FutureProcess(TaskCompletionSource<T> tcs)
-        {
-            _tcs = tcs;
-        }
-
-        public override void SendUserMessage(PID pid, object message, PID sender)
-        {
-            if (message is T)
-            {
-                _tcs.TrySetResult((T) message);
-                pid.Stop();
-            }
-        }
-
-        public override void SendSystemMessage(PID pid, object message)
-        {
-        }
-    }
+﻿// -----------------------------------------------------------------------
+//  <copyright file="Futures.cs" company="Asynkron HB">
+//      Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Proto
+{
+    public class FutureProcess<T> : Process
+    {
+        private readonly TaskCompletionSource<T> _tcs;
+        private readonly CancellationToken _cancellationToken;
+
+        public FutureProcess(TaskCompletionSource<T> tcs) : this(tcs, CancellationToken.None) { }
+        public FutureProcess(TaskCompletionSource<T> tcs, CancellationToken cancellationToken)
+        {
+            _tcs = tcs;
+            _cancellationToken = cancellationToken;
+        }
+
+        public override void SendUserMessage(PID pid, object message, PID sender)
+        {
+            if (message is T)
+            {
+                if (_cancellationToken.IsCancellationRequested) return;
```

Comment:
Well the timeout task has to know the PID value in order to stop() it, and currently the future process knows only about it when SendUserMessage is called ...
Could the FutureProcess do the PID/ProcessRegistry registration itself in its ctor ?